**Taller de limpieza y filtrado**

*1.) Importar un CSV desde Datos Abiertos*

In [ ]:
pip install sodapy

In [ ]:
import numpy as np
import pandas as pd
from sodapy import Socrata

In [71]:
# Unauthenticated client only works with public data sets. Note 'None'
# in place of application token, and no username or password:
client = Socrata("www.datos.gov.co", None)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(www.datos.gov.co,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

# First 2000 results, returned as JSON from API / converted to Python list of
# dictionaries by sodapy.
results = client.get("sdvb-4x4j", limit=4030)

# Convert to pandas DataFrame
asignacion_vacuna = pd.DataFrame.from_records(results)
convert_dict = {'cantidad': int}
asignacion_vacuna=asignacion_vacuna.astype(convert_dict)
asignacion_vacuna


num_resolucion  ...              fecha_corte
0               168  ...  2021-11-11T00:00:00.000
1               168  ...  2021-11-11T00:00:00.000
2               168  ...  2021-11-11T00:00:00.000
3               168  ...  2021-11-11T00:00:00.000
4               168  ...  2021-11-11T00:00:00.000
...             ...  ...                      ...
4025           1815  ...  2021-11-11T00:00:00.000
4026           1815  ...  2021-11-11T00:00:00.000
4027           1815  ...  2021-11-11T00:00:00.000
4028           1815  ...  2021-11-11T00:00:00.000
4029           1815  ...  2021-11-11T00:00:00.000

[4030 rows x 9 columns]

*2.) Limpiar datos*

**Efecto de la funcion isnull()**

In [ ]:
asignacion_vacuna.isnull()

num_resolucion  fecha_resolucion  ...  uso_vacuna  fecha_corte
0              False             False  ...       False        False
1              False             False  ...       False        False
2              False             False  ...       False        False
3              False             False  ...       False        False
4              False             False  ...       False        False
...              ...               ...  ...         ...          ...
4025           False             False  ...       False        False
4026           False             False  ...       False        False
4027           False             False  ...       False        False
4028           False             False  ...       False        False
4029           False             False  ...       False        False

[4030 rows x 9 columns]

Retorna True si el valor de la celda es NaN o None, False si tiene un valor válido





**Efecto de la funcion dropna(axis=0,how='any')**

In [ ]:
asignacion_vacuna.dropna(axis=0,how='any')

num_resolucion  ...              fecha_corte
0               168  ...  2021-11-11T00:00:00.000
1               168  ...  2021-11-11T00:00:00.000
2               168  ...  2021-11-11T00:00:00.000
3               168  ...  2021-11-11T00:00:00.000
4               168  ...  2021-11-11T00:00:00.000
...             ...  ...                      ...
4025           1815  ...  2021-11-11T00:00:00.000
4026           1815  ...  2021-11-11T00:00:00.000
4027           1815  ...  2021-11-11T00:00:00.000
4028           1815  ...  2021-11-11T00:00:00.000
4029           1815  ...  2021-11-11T00:00:00.000

[4029 rows x 9 columns]

Elimina filas en caso de que algun valor de esta sea NaN o None, en este caso solo elimino una fila, paso de tener 4030 a 4029

**Efecto de la funcion dropna(axis=0,how='all')**

In [ ]:
asignacion_vacuna.dropna(axis=0,how='all')

num_resolucion  ...              fecha_corte
0               168  ...  2021-11-11T00:00:00.000
1               168  ...  2021-11-11T00:00:00.000
2               168  ...  2021-11-11T00:00:00.000
3               168  ...  2021-11-11T00:00:00.000
4               168  ...  2021-11-11T00:00:00.000
...             ...  ...                      ...
4025           1815  ...  2021-11-11T00:00:00.000
4026           1815  ...  2021-11-11T00:00:00.000
4027           1815  ...  2021-11-11T00:00:00.000
4028           1815  ...  2021-11-11T00:00:00.000
4029           1815  ...  2021-11-11T00:00:00.000

[4030 rows x 9 columns]

Elimina filas en caso de que todos los valores de esta sea NaN o None, en este caso no se elimino ninguna fila

**Efecto de la funcion dropna(axis=1, how='any')**

In [ ]:
asignacion_vacuna.dropna(axis=1, how='any')

num_resolucion         fecha_resolucion  ... cantidad              fecha_corte
0               168  2021-02-16T00:00:00.000  ...     6570  2021-11-11T00:00:00.000
1               168  2021-02-16T00:00:00.000  ...      120  2021-11-11T00:00:00.000
2               168  2021-02-16T00:00:00.000  ...      564  2021-11-11T00:00:00.000
3               168  2021-02-16T00:00:00.000  ...     2556  2021-11-11T00:00:00.000
4               168  2021-02-16T00:00:00.000  ...    12582  2021-11-11T00:00:00.000
...             ...                      ...  ...      ...                      ...
4025           1815  2021-11-11T00:00:00.000  ...    25200  2021-11-11T00:00:00.000
4026           1815  2021-11-11T00:00:00.000  ...    20400  2021-11-11T00:00:00.000
4027           1815  2021-11-11T00:00:00.000  ...     6000  2021-11-11T00:00:00.000
4028           1815  2021-11-11T00:00:00.000  ...      200  2021-11-11T00:00:00.000
4029           1815  2021-11-11T00:00:00.000  ...     2400  2021-11-11T00:00:00.000

[4030 rows x 7 columns]

Elimina columnas en caso de que cualquier valor de la columna sea NaN o None, en este caso elimino 2 columnas

**Efecto de la funcion fillna(value=3)**

In [54]:
df = asignacion_vacuna[asignacion_vacuna.isna().any(axis=1)] 
df


num_resolucion  ...              fecha_corte
2401           1024  ...  2021-11-11T00:00:00.000

[1 rows x 9 columns]

In [57]:

df = asignacion_vacuna.fillna (value=3)



In [63]:
df.loc[2401]

num_resolucion                           1024
fecha_resolucion      2021-07-16T00:00:00.000
a_o                                      2022
cod_territorio                              3
nom_territorio                          CAUCA
laboratorio_vacuna                     PFIZER
cantidad                               -16380
uso_vacuna                                  3
fecha_corte           2021-11-11T00:00:00.000
Name: 2401, dtype: object

A los valores que esta nulos les asigna el valor que esta en el parametro

**Efecto de df.fillna(df.mean())**

In [64]:
asignacion_vacuna.mean()

num_resolucion             inf
a_o                        inf
cantidad          1.451692e+04
dtype: float64

*3.) Filtrado y consulta de datos*

**df['columna n’] >= valor**

In [45]:

asignacion_vacuna["cantidad"] >= 600

0        True
1       False
2       False
3        True
4        True
        ...  
4025     True
4026     True
4027     True
4028    False
4029     True
Name: cantidad, Length: 4030, dtype: bool

Retorna un valor booleano para cada fila, en este caso retorna True si el valor de cantidad es mayor o igual a 600